source/common/upstream/upstream_impl.h

```diff
@@ -977,7 +977,8 @@ class ClusterInfoImpl : public ClusterInfo,
 
   bool perEndpointStatsEnabled() const override { return per_endpoint_stats_; }
 
-  UpstreamLocalAddressSelectorConstSharedPtr getUpstreamLocalAddressSelector() const override {
+  const UpstreamLocalAddressSelectorConstSharedPtr
+  getUpstreamLocalAddressSelector() const override {
```

Comment:
---

This `const` make no sense because this method return a copy of the shared pointer.

---
And I actually feel a little head ache to this, I prefer to return a `const UpstreamLocalAddressSelector&` here. But I found the `HappyEyeballsConnectionProvider` want to share ownership of this selector.

We have lots of similar scenarios, in most case we needn't ownership, but in some position, we need ownership, finally, we spread the shared pointer becase the latter requirement.

But I have no a good idea to resolve this except to use `enable_share_from_this`.